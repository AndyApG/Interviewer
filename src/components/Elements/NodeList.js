import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Node, animation } from 'network-canvas-ui';
import StaggeredTransitionGroup from '../../utils/StaggeredTransitionGroup';
import { scrollable, droppable, draggable, selectable } from '../../behaviours';

const EnhancedNode = draggable(selectable(Node));

/**
  * Renders a list of Node.
  */
const NodeList = ({
  nodes,
  nodeColor,
  label,
  selected,
  handleSelectNode,
  handleDropNode,
  draggableType,
  hover,
  isDragging,
}) => {
  const classNames = cx(
    'node-list',
    { 'node-list--hover': hover },
    { 'node-list--drag': isDragging },
  );

  return (
    <StaggeredTransitionGroup
      className={classNames}
      component="div"
      delay={animation.duration.fast * 0.2}
      duration={animation.duration.slow}
      start={animation.duration.slow + 3}
      transitionName="node-list--transition"
      transitionLeave={false}
    >
      { hover &&
        <Node key="placeholder" placeholder />
      }
      {
        nodes.map(node => (
          <span key={node.uid}>
            <EnhancedNode
              color={nodeColor}
              label={label(node)}
              selected={selected(node)}
              onSelected={() => handleSelectNode(node)}
              onDropped={hits => handleDropNode(hits, node)}
              draggableType={draggableType}
              {...node}
            />
          </span>
        ))
      }
    </StaggeredTransitionGroup>
  );
};

NodeList.propTypes = {
  nodes: PropTypes.array.isRequired,
  nodeColor: PropTypes.string,
  handleSelectNode: PropTypes.func,
  handleDropNode: PropTypes.func,
  label: PropTypes.func,
  selected: PropTypes.func,
  draggableType: PropTypes.string,
  hover: PropTypes.bool,
  isDragging: PropTypes.bool.isRequired,
};

NodeList.defaultProps = {
  nodes: [],
  nodeColor: '',
  label: () => (''),
  selected: () => false,
  handleSelectNode: () => {},
  handleDropNode: () => {},
  draggableType: '',
  hover: false,
};

function mapStateToProps(state, ownProps) {
  return {
    isDragging: state.draggable.isDragging &&
      state.draggable.draggableType === ownProps.acceptsDraggableType,
  };
}

export default connect(mapStateToProps)(droppable(scrollable(NodeList)));
