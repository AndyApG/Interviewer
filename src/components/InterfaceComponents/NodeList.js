import React, { Component } from 'react';
import Node from './Node';

class NodeList extends Component {
  render() {
    const {
      network
    } = this.props;

    console.log(network);

    return (
      <div className='nodelist'>
      { network.nodes.map((node, index) => {
        return <Node key={ index } { ...node } />;
      }) }
      </div>
    );
  }
}

export default NodeList;
