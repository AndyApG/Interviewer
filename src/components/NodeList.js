import React, { Component } from 'react';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { find, get, isEqual } from 'lodash';
import cx from 'classnames';
import Color from 'color';
import { Node, animation, colorDictionary } from 'network-canvas-ui';
import { TransitionGroup } from 'react-transition-group';
import { Node as NodeTransition } from './Transition';
import { scrollable, selectable } from '../behaviours';
import {
  DragSource,
  DropTarget,
  MonitorDropTarget,
  MonitorDragSource,
} from '../behaviours/DragAndDrop';

const EnhancedNode = DragSource(selectable(Node));

/**
  * Renders a list of Node.
  */
class NodeList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      nodes: props.nodes,
      stagger: true,
      exit: true,
    };

    this.refreshTimer = null;
  }

  componentWillReceiveProps(newProps) {
    // Don't update if nodes are the same
    if (isEqual(newProps.nodes, this.state.nodes)) {
      return;
    }

    // if we provided the same id, then just update normally
    if (newProps.listId === this.props.listId) {
      this.setState({ exit: false }, () => {
        this.setState({ nodes: newProps.nodes, stagger: false });
      });
      return;
    }

    // Otherwise, transition out and in again
    this.props.scrollTop(0);

    this.setState({ exit: true }, () => {
      this.setState(
        { nodes: [], stagger: true },
        () => {
          if (this.refreshTimer) { clearTimeout(this.refreshTimer); }
          this.refreshTimer = setTimeout(
            () => this.setState({
              nodes: newProps.nodes,
              stagger: true,
            }),
            animation.duration.slow,
          );
        },
      );
    });
  }

  render() {
    const {
      nodeColor,
      label,
      selected,
      onSelect,
      itemType,
      isOver,
      willAccept,
      meta,
    } = this.props;

    const {
      stagger,
      nodes,
      exit,
    } = this.state;

    const isSource = !!find(nodes, ['uid', get(meta, 'uid', null)]);
    const isValidTarget = !isSource && willAccept;
    const isHovering = isValidTarget && isOver;

    const classNames = cx(
      'node-list',
      { 'node-list--drag': isValidTarget },
    );

    const backgroundColor = Color(
      nodeColor || colorDictionary['node-base'],
    ).alpha(0.1);

    const styles = isHovering ? { backgroundColor } : {};

    return (
      <TransitionGroup
        className={classNames}
        style={styles}
        exit={exit}
      >
        {
          nodes.map((node, index) => (
            <NodeTransition
              key={`${node.uid}`}
              index={index}
              stagger={stagger}
            >
              <EnhancedNode
                color={nodeColor}
                label={`${label(node)}`}
                selected={selected(node)}
                onSelected={() => onSelect(node)}
                meta={() => ({ ...node, itemType })}
                {...node}
              />
            </NodeTransition>
          ))
        }
      </TransitionGroup>
    );
  }
}

NodeList.propTypes = {
  nodes: PropTypes.array.isRequired,
  nodeColor: PropTypes.string,
  onSelect: PropTypes.func,
  itemType: PropTypes.string,
  label: PropTypes.func,
  selected: PropTypes.func,
  isOver: PropTypes.bool,
  willAccept: PropTypes.bool,
  meta: PropTypes.object,
  listId: PropTypes.string.isRequired,
  scrollTop: PropTypes.func,
};

NodeList.defaultProps = {
  nodes: [],
  nodeColor: '',
  label: () => (''),
  selected: () => false,
  onSelect: () => {},
  onDrop: () => {},
  itemType: 'NODE',
  isOver: false,
  willAccept: false,
  isDragging: false,
  meta: {},
  scrollTop: () => {},
};

export default compose(
  DropTarget,
  MonitorDropTarget(['isOver', 'willAccept']),
  MonitorDragSource(['meta', 'isDragging']),
  scrollable,
)(NodeList);
