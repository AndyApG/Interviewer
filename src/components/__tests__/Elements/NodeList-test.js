/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import NodeList from '../../Elements/NodeList';

describe('NodeList component', () => {
  it('renders ok', () => {
    const component = shallow(
      <NodeList>
        <span>foo</span>
      </NodeList>
    );

    expect(component).toMatchSnapshot();
  });

});
