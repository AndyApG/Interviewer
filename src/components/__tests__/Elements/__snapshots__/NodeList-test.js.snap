exports[`NodeList component renders ok 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Droppable)
    name="nodelist">
    <div
        className="node-list">
        <Node
            label="undefined" />
    </div>
</Connect(Droppable)>,
  "nodes": Array [
    <Connect(Droppable)
      name="nodelist">
      <div
            className="node-list">
            <Node
                  label="undefined" />
      </div>
</Connect(Droppable)>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <NodeList
        network={
                Object {
                        "nodes": Array [
                          Object {
                            "name": "test",
                          },
                        ],
                      }
        } />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": NodeList {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "network": Object {
            "nodes": Array [
              Object {
                "name": "test",
              },
            ],
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Droppable)
          name="nodelist">
          <div
                    className="node-list">
                    <Node
                              label="undefined" />
          </div>
</Connect(Droppable)>,
        "_debugID": 2,
        "_renderedOutput": <Connect(Droppable)
          name="nodelist">
          <div
                    className="node-list">
                    <Node
                              label="undefined" />
          </div>
</Connect(Droppable)>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <NodeList
    network={
        Object {
            "nodes": Array [
              Object {
                "name": "test",
              },
            ],
          }
    } />,
}
`;
